site/docs/spark.md

```diff
@@ -519,6 +519,59 @@ data.writeTo("prod.db.table")
     .createOrReplace()
 ```
 
+## Writing against partitioned table
+
+Iceberg requires the data to be sorted according to the partition spec in prior to write against partitioned table.
```

Comment:
This is a little to specific because Iceberg actually just requires that your data is clustered by partition in each task. Maybe this is the easiest way to get the idea across initially, but it would be nice to have a note box that explains that both global and local sorts will work.

Reply:
Please correct me if I'm missing something. (Sorry for being too correct technically - I'm also a learner of Iceberg so just to understand correctly.)

If I understand correctly, at least "Iceberg Spark writer" requires the data to be sorted according to the partition spec in task (Spark partition), not just the data to be clustered by partition.


Below query fails:

```
spark.sql("""
CREATE TABLE iceberg_catalog.default.sample1 (
    id bigint,
    data string,
    category string)
USING iceberg
PARTITIONED BY (category)
""")

val data = (0 to 100000).map { id =>
  (id, s"hello$id", s"category-${id % 100}")
}

data.toDF("id", "data", "category").repartition(100, col("category")).sortWithinPartitions("id").writeTo("iceberg_catalog.default.sample1").append()
```

Mentioning the global and local sorts would be nice to have. Thanks! Will add.

Reply:
You're right that the query fails. Although the data is partitioned to group by category, Spark will use hash partitioning so there is no guarantee that each partition contains only one category. If it there were, it would work. Some partitions are going to contain more than one category, and a partition is a task. Sorting by id mixes the categories together, which causes the failure. I think it would also fail without the sort because the data coming from the map side of the shuffle would also be mixed together.

For repartitioning, it appears like Spark will cluster the data, but you actually need a sort to do it. However, there are other cases where you don't need to add an explicit sort, like when you can guarantee that only one partition is written by a job (writing an agg table for one day) or when the input data already aligns with the partitioning of the output.

Reply:
Yeah you're right that repartition wouldn't make each partition to have only one category, and I think that's normal as we probably don't want to have large number of partitions (tasks) due to the unpredictable cardinality of category.

I understand there're cases where we don't need to add an explicit sort. Probably they would become the cases no implicit sort is performed as well if Spark knows enough information on data alignment & required alignment - it'd be ideal SPARK-23889 would also cover the case.